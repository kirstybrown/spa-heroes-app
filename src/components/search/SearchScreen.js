import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';

import React from 'react'
import { HeroCard } from '../hero/HeroCard';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const {q = '' } = queryString.parse(location.search);

  const [ formValues, handleInputChange ] = useForm( {
    searchText: q,
  } );

  const { searchText } = formValues;

  const heroesFiltered = getHeroesByName(q);

  const handleSearch = (e) => {
    e.preventDefault(); 
    navigate(`?q=${ searchText }`)
  }

  return (
    <>
        <h1>Search</h1>
        <hr />

        <div className='row'>
          <div className='col-5'>
            <h4>Search</h4>
            <hr />

            <form onSubmit={ handleSearch }>
              <input 
                type='text'
                placeholder='Search for a hero'
                className='form-control'
                name='searchText'
                autoComplete='off'
                value={ searchText }
                onChange={handleInputChange}
              />


              <button
                className='btn btn-outline-primary mt-1'
                type='submit'
                onClick={ handleSearch }
              >
                Search...
              </button>
            </form>

          </div>

          <div className='col-7'>
            <h4>Results</h4>
            <hr />

            {
              (q === '')
                  ? <div className='alert alert-info'> Please enter a hero!</div>
                  : ( heroesFiltered.length === 0 )
                    && <div className='alert alert-danger'> No results for "{q}", sorry! </div>
            }

            {
              heroesFiltered.map(hero => (
                <HeroCard 
                  key={ hero.id }
                  { ...hero }
                />
              ))
            }


          </div>
        </div>
    </>
  )
}
