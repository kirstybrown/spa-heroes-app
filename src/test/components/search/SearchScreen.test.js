import { mount } from "enzyme"
import { MemoryRouter, useNavigate } from "react-router-dom"
import { SearchScreen } from "../../../components/search/SearchScreen"

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));


describe('Tests for <SearchScreen /> component', () => { 

    test('should show correctly with default values', () => { 

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );        
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim()).toBe('Please enter a hero!');
     });

    test('should show Batman and the input with the value of queryString', () => { 

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        );  
        expect( wrapper.find('input').prop('value')).toBe('batman');

      });

    test('should show an error if no hero found', () => { 

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen />
            </MemoryRouter>
        );  
        expect( wrapper.find('.alert-danger').text().trim() ).toBe('No results for "batman123", sorry!');

       });

    test('should call the navigate to the new screen', () => { 

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen />
            </MemoryRouter>
        );  
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        })

        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');
    })

 })