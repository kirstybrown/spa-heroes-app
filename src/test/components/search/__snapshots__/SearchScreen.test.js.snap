// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <SearchScreen /> component should show correctly with default values 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search for a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          onClick={[Function]}
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
         Please enter a hero!
      </div>
    </div>
  </div>,
]
`;
